05-java-mqtt/src/main/java/demo/MqttNetworkManager.java

```diff
@@ -0,0 +1,163 @@
+     * @throws MqttException if some MQTT related error occur.
+     * @return the token to track the listen asynchronous operation.
+     */
+    public IMqttToken listen(final String topic) throws MqttException {
+        final String broker = "tcp://" + this.address + ":" + this.port;
+        final MqttClientPersistence persistence = new MemoryPersistence();
+        this.mqttClient = new MqttAsyncClient(broker, this.clientId, persistence);
+        mqttClient.connect().waitForCompletion();
+        return mqttClient.subscribe(topic, this.qos, null, null, this::handleMessage);
+    }
+
+    /**
+     * Shutdowns the MQTT client.
+     * @return the token to track the asynchronous task
+     * @throws MqttException if some MQTT related error occur
+     */
+    public IMqttToken stop() throws MqttException {
+        return this.mqttClient.disconnect();
+    }
+
+    @SuppressWarnings("PMD.UnusedFormalParameter")
+    private void handleMessage(final String topic, final MqttMessage message) throws IOException, ClassNotFoundException {
+        final Object received = new ObjectInputStream(new ByteArrayInputStream(message.getPayload()))
+                .readObject();
+        if (received instanceof Map) {
+            ((Map) received).forEach((src, msg) -> receiveMessage((DeviceUID) src, (Map<CodePath, Object>) msg));
+        }
+    }
+
+    private void receiveMessage(final DeviceUID src, final Map<CodePath, Object> msg) {
+        messages.put(src, msg);
+    }
+
+    /**
+     * Called by ProtelisVM read the stored messages.
+     * @return the currently stored messages
+     */
+    @Override
+    public Map<DeviceUID, Map<CodePath, Object>> getNeighborState() {
+        final Map<DeviceUID, Map<CodePath, Object>> t = messages;
+        messages = new HashMap<>();
+        return t;
+    }
+
+    /**
+     * Called by ProtelisVM to send a message to the neighbors.
+     * @param toSend the message to be sent.
+     */
+    @Override
+    public void shareState(final Map<CodePath, Object> toSend) {
+        final Map<DeviceUID, Map<CodePath, Object>> msg = Stream.of(
+                new AbstractMap.SimpleImmutableEntry<>(deviceUID, toSend)
+        ).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        final ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        ObjectOutput out = null;
+        try {
+            out = new ObjectOutputStream(bos);
+            out.writeObject(msg);
+            out.flush();
+            final MqttMessage message = new MqttMessage(bos.toByteArray());
+            neighbors.forEach(publish(message));
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private Consumer<String> publish(final MqttMessage message) {
+        return topic -> {
+            try {
+                mqttClient.publish(topic, message).waitForCompletion();
+            } catch (MqttException e) {
+                e.printStackTrace();
```

Comment:
rethrow it in a runtime exception

Reply:
Pmd gives me back an "Avoid throwing raw exception types" error. How should I proceed?

Reply:
Use a more specific runtime exception :)
IllegalStateException?